Signal/src/views/PinEntryView.m

```diff
@@ -0,0 +1,205 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+#import "PinEntryView.h"
+#import "Signal-Swift.h"
+#import <SignalMessaging/UIColor+OWS.h>
+#import <SignalMessaging/UIFont+OWS.h>
+#import <SignalMessaging/UIView+OWS.h>
+#import <SignalMessaging/ViewControllerUtils.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface PinEntryView () <UITextFieldDelegate>
+
+@property (nonatomic) UITextField *pinTextfield;
+@property (nonatomic) OWSFlatButton *submitButton;
+@property (nonatomic) UILabel *instructionsLabel;
+
+@end
+
+@implementation PinEntryView : UIView
+
+- (instancetype)initWithFrame:(CGRect)frame
+{
+    self = [super initWithFrame:frame];
+    if (!self) {
+        return self;
+    }
+
+    [self createContents];
+
+    return self;
+}
+
+#pragma mark - view creation
+
+- (UILabel *)createLabelWithText:(nullable NSString *)text
+{
+    UILabel *label = [UILabel new];
+    label.textColor = [UIColor blackColor];
+    label.text = text;
+    label.font = [UIFont ows_regularFontWithSize:ScaleFromIPhone5To7Plus(14.f, 16.f)];
+    label.numberOfLines = 0;
+    label.lineBreakMode = NSLineBreakByWordWrapping;
+    label.textAlignment = NSTextAlignmentCenter;
+    [self addSubview:label];
+    return label;
+}
+
+- (void)createPinTextfield
+{
+    self.pinTextfield = [UITextField new];
+    self.pinTextfield.textColor = [UIColor blackColor];
+    self.pinTextfield.placeholder
+        = NSLocalizedString(@"2FA_PIN_DEFAULT_TEXT", @"Text field placeholder when entering a 'two-factor auth pin'.");
+    self.pinTextfield.font = [UIFont ows_mediumFontWithSize:ScaleFromIPhone5To7Plus(30.f, 36.f)];
+    self.pinTextfield.textAlignment = NSTextAlignmentCenter;
+    self.pinTextfield.keyboardType = UIKeyboardTypeNumberPad;
+    self.pinTextfield.delegate = self;
+    self.pinTextfield.secureTextEntry = YES;
+    self.pinTextfield.textAlignment = NSTextAlignmentCenter;
+    [self addSubview:self.pinTextfield];
+}
+
+- (UILabel *)createForgotLink
+{
+    UILabel *label = [UILabel new];
+    label.textColor = [UIColor ows_materialBlueColor];
+    NSString *text = NSLocalizedString(
+        @"REGISTER_2FA_FORGOT_PIN", @"Label for 'I forgot my PIN' link in the 2FA registration view.");
```

Comment:
I suspect we'll want different "forgot" copy in the registration w/ pin view (where the stakes are high) v. the reminder view (where the stakes are low).  They have different resolutions too: in the former, you wait.  In the latter, we might want to offer to clear the PIN and/or let users select a new PIN if they forgot it.

Reply:
I agree that the *behavior* of tapping "I forgot my PIN" should be different depending on context (that's achieved via the delegate method `pinEntryViewForgotPinLinkTapped`), but the actual text of the button which prompts that behavior: "I forgot my PIN" is the same in both places, which seems reasonable to me.